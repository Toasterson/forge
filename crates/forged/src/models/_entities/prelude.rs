//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.1

pub use super::gates::Entity as Gates;
pub use super::notes::Entity as Notes;
pub use super::users::Entity as Users;
