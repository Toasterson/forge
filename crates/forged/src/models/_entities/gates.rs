//! `SeaORM` Entity, @generated by sea-orm-codegen 1.0.1

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "gates")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    #[sea_orm(unique)]
    pub pid: Uuid,
    #[sea_orm(unique)]
    pub name: String,
    pub version: String,
    pub branch: String,
    #[sea_orm(column_type = "JsonBinary", nullable)]
    pub transforms: Option<Json>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
