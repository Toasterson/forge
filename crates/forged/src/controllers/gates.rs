#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use crate::models::_entities::gates::{ActiveModel, Column, Entity, Model};
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AddParams {
    pub name: String,
    pub version: String,
    pub id: Uuid,
    pub branch: String,
    pub transforms: Option<Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateParams {
    pub version: Option<String>,
    pub branch: Option<String>,
    pub transforms: Option<Vec<Value>>,
}

impl UpdateParams {
    fn update_model(&self, item: &mut ActiveModel) {
        if let Some(version) = &self.version {
            item.version = ActiveValue::Set(version.clone());
        }
        if let Some(branch) = &self.branch {
            item.branch = ActiveValue::Set(branch.clone());
        }
        if let Some(transforms) = &self.transforms {
            if transforms.is_empty() {
                item.transforms = ActiveValue::NotSet;
            } else {
                let tr = serde_json::to_value(transforms.clone()).unwrap();
                item.transforms = ActiveValue::Set(Some(tr));
            }
        }
    }
}

pub async fn list(State(ctx): State<AppContext>) -> Result<Response> {
    let gates = Entity::find().all(&ctx.db).await?;
    format::json(&gates)
}

async fn load_gate(ctx: &AppContext, name: &str) -> Result<Model> {
    let gate = Entity::find()
        .filter(Column::Name.contains(name))
        .one(&ctx.db)
        .await?;
    gate.ok_or_else(|| Error::NotFound)
}

pub async fn add(State(ctx): State<AppContext>, Json(params): Json<AddParams>) -> Result<Response> {
    let gate = ActiveModel {
        pid: ActiveValue::Set(params.id),
        name: ActiveValue::Set(params.name),
        version: ActiveValue::Set(params.version),
        branch: ActiveValue::Set(params.branch),
        transforms: ActiveValue::Set(params.transforms),
        ..Default::default()
    };

    let gate = gate.insert(&ctx.db).await?;
    format::json(&gate)
}

pub async fn update(
    State(ctx): State<AppContext>,
    Path(name): Path<String>,
    Json(params): Json<UpdateParams>,
) -> Result<Response> {
    let gate = load_gate(&ctx, &name).await?;
    let mut gate = gate.into_active_model();
    params.update_model(&mut gate);
    let gate = gate.update(&ctx.db).await?;
    format::json(&gate)
}

pub async fn get_one(State(ctx): State<AppContext>, Path(name): Path<String>) -> Result<Response> {
    let gate = load_gate(&ctx, &name).await?;
    format::json(&gate)
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("gates/")
        .add("/", get(list))
        .add("/", post(add))
        .add("/:name", post(update))
        .add("/:name", get(get_one))
}
