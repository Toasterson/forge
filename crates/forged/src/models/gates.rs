use super::_entities::gates::{ActiveModel, Column, Entity, Model};
use loco_rs::prelude::*;

impl ActiveModelBehavior for ActiveModel {
    // extend activemodel below (keep comment for generators)
}

impl Model {
    /// finds a gate by the provided name
    ///
    /// # Errors
    ///
    /// When could not find gate by the given name or DB query error
    pub async fn find_by_name<S: AsRef<str>>(
        db: &DatabaseConnection,
        name: S,
    ) -> ModelResult<Self> {
        let gate = Entity::find()
            .filter(query::condition().eq(Column::Name, name.as_ref()).build())
            .one(db)
            .await?;
        gate.ok_or_else(|| ModelError::EntityNotFound)
    }
}
