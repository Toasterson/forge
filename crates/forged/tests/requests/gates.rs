use insta::{assert_debug_snapshot, with_settings};
use forged::app::App;
use loco_rs::testing;
use serial_test::serial;
use forged::models::_entities::gates;

// TODO: see how to dedup / extract this to app-local test utils
// not to framework, because that would require a runtime dep on insta
macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("gate_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_add() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        let payload = serde_json::json!({
            "id": "a1a6a18e-61bf-4d13-b00b-ff543c91e890",
            "name": "userland",
            "version": "0.5.11",
            "branch": "2024.0.0"
        });

        let _response = request.post("/api/gates").json(&payload).await;
        let saved_gate = gates::Model::find_by_name(&ctx.db, "userland").await;

        with_settings!({
            filters => testing::cleanup_user_model()
        }, {
            assert_debug_snapshot!(saved_gate);
        });
    })
        .await;
}