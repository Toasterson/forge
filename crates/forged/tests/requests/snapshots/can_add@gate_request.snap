---
source: crates/forged/tests/requests/gates.rs
assertion_line: 37
expression: saved_gate
---
Ok(
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        pid: PID,
        name: "userland",
        version: "TOKEN",
        branch: "TOKEN",
        transforms: None,
    },
)
